### ISSUE TITLE
Type-safe way to create EntityReference

### ISSUE THREAD

**Issue by zaha** [delegateas/XrmDefinitelyTyped#38]
Hi everyone,

unless I missed something there is currently no type-safe way to create an EntityReference. This leaves room for errors like this:

```js
Xrm.Page.getAttribute("new_language").setValue([{ id: languageId, entityType: "tv_langauge" }]);
```
(simple spelling mistake with two switched up letters: lang**au**ge =>lang**ua**ge )

I can think of two better way of handling this:

1.  Introduce some sort of factory method that takes an EntityType as parameter:
```js
let entityRef = Xrm.EntityReference.Create<New.EntityType>(id);
```
2. Provide a constant for the type itself:
```js
let entityRef = { id: id, entityType: New.Language.EntityType }
```

Would this be possible or feasible?

**Comment by mktange** [comment]
I think a better approach is to just make the `EntityReference` interface generic, such that it can be customized per attribute which string literals are allowed as entity logical names.

It would involve changing this:

https://github.com/delegateas/XrmDefinitelyTyped/blob/f9361ff43316bfc7e4911edead78f900410f7de8/src/XrmDefinitelyTyped/Resources/xrm.d.ts#L53-L57

to this:

```typescript
    interface EntityReference<T> {
        id: string;
        entityType: T;
        name?: string | null;
    }
```

And then have every generated interface that uses `EntityReference`, use the correct generic version instead. So in your example the type definition for that attribute would in the end be something like:

```typescript
   getAttribute("new_language"): Attribute<EntityReference<"tv_language">[]>;
   getAttribute("ownerid"): Attribute<EntityReference<"systemuser" | "team">[]>;
```

This would overall make the typings much stronger and more reliable. 

However, in some cases it may introduce some complications for developers. I.e. when a function/API call returns a entity logical name as a plain `string` type, and thus you are not able use it for that field without type-casting it first. Not sure how big of an issue that would be though.

**Comment by zaha** [comment]
That would be an even better solution. For me personally I would gladly trade the auto-complete/type-safety for some casting in those special cases, but I can't speak for other people.

At the very least, having some auto-generated constants for the entity types would be nicer than having to manually type those strings every time.

**Comment by Flydancer** [comment]
> At the very least, having some auto-generated constants for the entity types would be nicer than having to manually type those strings every time.

Yes, constants for Fieldnames and/or Entitytypes would be great!

**Comment by henrikhannemose** [comment]
This is a nice change, but quite breaking. I get errors when updating to 2.7.0 in projects where I use e.g. `Xrm.LookupAttribute` (which should of course just be changed to the newer and better syntax). 

I have unlisted 2.7.0 again, and @Tolini will re-publish it as version 3.0 to signal the breaking changes (along with a fix for #67?)
